### DISCUSSION TITLE
Get files by its url from storage.

### DISCUSSION THREAD

**Discussion by usamashafiqs** [appwrite/appwrite#3474] [Q&A] [answered]
How to get files by its url from storage?

**Comment by stnguyen90** [comment, marked as answer]
This depends on whether you want to view the data inline or download the file.

To view a file, send a GET request to `/v1/storage/buckets/{bucketId}/files/{fileId}/view`. Refer to the [Get File for View docs](https://appwrite.io/docs/client/storage?sdk=web-default#storageGetFileView) for more info.

To download a file, send a GET request to `/v1/storage/buckets/{bucketId}/files/{fileId}/download`. Refer to the [Get File for Download docs](https://appwrite.io/docs/client/storage?sdk=web-default#storageGetFileDownload) for more info.

**Reply by usamashafiqs** [reply]
Both endpoint return ``Uint8List bytes`` in flutter client.
then we use [Image.memory()](https://api.flutter.dev/flutter/widgets/Image/Image.memory.html) widget to encode bytes to image formats: JPEG, PNG, GIF etc.

**But in some cases we need ``Direct URL`` of images or files.**

Like firebase : [Firebase file](https://firebasestorage.googleapis.com/v0/b/wood-tag.appspot.com/o/images%2Fproduct%2FIibn7qd3Wcs1SY7kCxKV1656567027765?alt=media&token=142c15da-fb12-4bc9-afc9-1f2dd75be9bd)
Firebase download URL endpoint: [Docs](https://firebase.google.com/docs/storage/flutter/download-files#download_data_via_url) 

Appwrite Example:
final Uint8List image = await ApiProvider.storage.getFilePreview(
        bucketId: "xxxxxxxx",
        fileId: "xxxxxxxxxxxxx",
     );
Image.memory(image);

**Reply by stnguyen90** [reply]
Sorry, I wasn't suggesting to use the SDK. You can create a string like `/v1/storage/buckets/{bucketId}/files/{fileId}/view` and that will be your URL.

**Reply by usamashafiqs** [reply]
How to add project id in this URL?

**Reply by stnguyen90** [reply]
Oh right! You can append `?project={projectId}`. The difficult part will be authentication. If the file isn't public, you'll need to somehow include a cookie or generate a JWT token and set that in the `x-appwrite-jwt` header.

**Reply by usamashafiqs** [reply]
Can you provide repo or example of implementation with authentication.

**Reply by stnguyen90** [reply]
Sorry I don't have any repo. You can call [`createJWT()`](https://appwrite.io/docs/client/account?sdk=flutter-default#accountCreateJWT) to get a JWT token and then pass that into the header. Something like:

```dart
Image.network(
  'https://{hostname}/v1/storage/buckets/{bucketId}/files/{fileId}/view?project={projectId}', 
  headers: {"x-appwrite-jwt": jwt}
)
```

**Reply by usamashafiqs** [reply]
Thanks! steven.

**Reply by northern-mindset** [reply]
Is it possible to address a file using its name? Something like `/v1/storage/buckets/{bucketId}/files/{path/to/fileName}/download`. I'm migrating from GCP Storage, where the file name (path, actually) is its unique ID.

**Reply by shibakar** [reply]
> ```dart
>  headers: {"x-appwrite-jwt": jwt}
> ```

jwt expires in 15 min right so need to create new one every 15 min

**Reply by hortigado** [reply]
@shibakar i get the cookie from clientio and not need jwt
